File: ts-sdk/client/src/gpa/tickArray.ts
import type { Account, Address, GetProgramAccountsApi, Rpc } from "@solana/kit";
import type { DynamicTickArrayFilter } from "./dynamicTickArray";
import {
  dynamicTickArrayStartTickIndexFilter,
  dynamicTickArrayWhirlpoolFilter,
  fetchAllDynamicTickArrayWithFilter,
} from "./dynamicTickArray";
import type { TickArray } from "../state/tickArray";
import { consolidateTickArray } from "../state/tickArray";
import type { FixedTickArrayFilter } from "./fixedTickArray";
import {
  fetchAllFixedTickArrayWithFilter,
  fixedTickArrayStartTickIndexFilter,
  fixedTickArrayWhirlpoolFilter,
} from "./fixedTickArray";

export type TickArrayFilter = {
  fixed: FixedTickArrayFilter;
  dynamic: DynamicTickArrayFilter;
  readonly __kind: unique symbol;
};

export function tickArrayStartTickIndexFilter(
  startTickIndex: number,
): TickArrayFilter {
  return {
    fixed: fixedTickArrayStartTickIndexFilter(startTickIndex),
    dynamic: dynamicTickArrayStartTickIndexFilter(startTickIndex),
  } as TickArrayFilter;
}

export function tickArrayWhirlpoolFilter(address: Address): TickArrayFilter {
  return {
    fixed: fixedTickArrayWhirlpoolFilter(address),
    dynamic: dynamicTickArrayWhirlpoolFilter(address),
  } as TickArrayFilter;
}

export async function fetchAllTickArrayWithFilter(
  rpc: Rpc<GetProgramAccountsApi>,
  ...filters: TickArrayFilter[]
): Promise<Account<TickArray>[]> {
  const fixedAccounts = await fetchAllFixedTickArrayWithFilter(
    rpc,
    ...filters.map((filter) => filter.fixed),
  );
  const dynamicAccounts = await fetchAllDynamicTickArrayWithFilter(
    rpc,
    ...filters.map((filter) => filter.dynamic),
  );

  const tickArrays: Account<TickArray>[] = [];
  for (const account of fixedAccounts) {
    tickArrays.push(consolidateTickArray(account));
  }
  for (const account of dynamicAccounts) {
    tickArrays.push(consolidateTickArray(account));
  }

  return tickArrays;
}

File: rust-sdk/client/src/gpa/tick_array.rs
use std::error::Error;

use solana_client::nonblocking::rpc_client::RpcClient;
use solana_pubkey::Pubkey;

use crate::{DecodedAccount, TickArray};

use super::{fetch_all_dynamic_tick_array_with_filter, DynamicTickArrayFilter};
use super::{fetch_all_fixed_tick_array_with_filter, FixedTickArrayFilter};

#[derive(Debug, Clone)]
pub enum TickArrayFilter {
    Whirlpool(Pubkey),
    StartTickIndex(i32),
}

impl From<TickArrayFilter> for FixedTickArrayFilter {
    fn from(val: TickArrayFilter) -> Self {
        match val {
            TickArrayFilter::Whirlpool(address) => FixedTickArrayFilter::Whirlpool(address),
            TickArrayFilter::StartTickIndex(tick_index) => {
                FixedTickArrayFilter::StartTickIndex(tick_index)
            }
        }
    }
}

impl From<TickArrayFilter> for DynamicTickArrayFilter {
    fn from(val: TickArrayFilter) -> Self {
        match val {
            TickArrayFilter::Whirlpool(address) => DynamicTickArrayFilter::Whirlpool(address),
            TickArrayFilter::StartTickIndex(tick_index) => {
                DynamicTickArrayFilter::StartTickIndex(tick_index)
            }
        }
    }
}

pub async fn fetch_all_tick_array_with_filter(
    rpc: &RpcClient,
    filters: Vec<TickArrayFilter>,
) -> Result<Vec<DecodedAccount<TickArray>>, Box<dyn Error>> {
    let fixed_filters = filters
        .clone()
        .into_iter()
        .map(|filter| filter.into())
        .collect();
    let fixed_tick_arrays = fetch_all_fixed_tick_array_with_filter(rpc, fixed_filters).await?;

    let dynamic_filters = filters
        .clone()
        .into_iter()
        .map(|filter| filter.into())
        .collect();
    let dynamic_tick_arrays =
        fetch_all_dynamic_tick_array_with_filter(rpc, dynamic_filters).await?;

    let mut tick_arrays: Vec<DecodedAccount<TickArray>> = Vec::new();

    for fixed_tick_array in fixed_tick_arrays {
        tick_arrays.push(DecodedAccount {
            address: fixed_tick_array.address,
            account: fixed_tick_array.account,
            data: TickArray::FixedTickArray(fixed_tick_array.data),
        });
    }

    for dynamic_tick_array in dynamic_tick_arrays {
        tick_arrays.push(DecodedAccount {
            address: dynamic_tick_array.address,
            account: dynamic_tick_array.account,
            data: TickArray::DynamicTickArray(dynamic_tick_array.data),
        });
    }

    Ok(tick_arrays)
}
